### PR TITLE
11.0 base dav - upload/delete

### PR DESCRIPTION
**Pull request by fkantelberg**
This should add upload and delete for the calendar type. I currently don't know how to handle the other types and have to test it more. Furthermore there is the fix for the utc problem on datetime.

### CODE REVIEW

File: base_dav/models/dav_collection.py [hunk 1]
```diff
@@ -81,6 +84,45 @@ def eval(self):
...
+        if self.dav_type == 'calendar':
+            if item.name != 'VCALENDAR':
+                return None
+            if not hasattr(item, 'vevent'):
+                return None
+            item = item.vevent
+        elif self.dav_type == 'addressbook' and item.name != 'VCARD':
+            return None
+
+        childs = {c.name.lower(): c for c in item.getChildren()}
```
Comment (hbrunn) [inline base_dav/models/dav_collection.py]: make this `children`

File: base_dav/models/dav_collection.py [hunk 2]
```diff
@@ -117,17 +159,35 @@ def to_vobject(self, record):
             )
         return result
 
+    @api.model
+    def _from_vobject_datetime(self, item):
+        if isinstance(item.value, (datetime, date)):
+            return item.value.strftime(tools.DEFAULT_SERVER_DATETIME_FORMAT)
```
Comment (hbrunn) [inline base_dav/models/dav_collection.py:168]: my guess would be that here the value can carry a timezone, and that you lose in this conversion. You probably need something like

```python
if myval.tzinfo:
    my_val = item.value.astimezone(pytz.utc)
```

File: base_dav/models/dav_collection.py [hunk 3]
```diff
@@ -117,17 +159,35 @@ def to_vobject(self, record):
...
+    def _from_vobject_char_n(self, item):
+        return item.family
+
     @api.multi
     def _to_vobject_datetime(self, record, field_name):
         result = fields.Datetime.from_string(record[field_name])
-        # TODO: this still generates wrong times
-        result.replace(tzinfo=vobject.icalendar.utc)
-        return result
+        return result.replace(tzinfo=tz.UTC)
```
Comment (hbrunn) [inline base_dav/models/dav_collection.py:191]: indeed, thanks!

File: base_dav/models/dav_collection.py [hunk 4]
```diff
@@ -117,17 +159,35 @@ def to_vobject(self, record):
...
 
     @api.multi
     def _to_vobject_datetime_rev(self, record, field_name):
         return record[field_name] and record[field_name]\
-            .replace('-', '').replace(' ', 'T').replace(':', '') + 'Z'
+            .replace('-', '').replace(' ', 'T').replace(':', '')
```
Comment (hbrunn) [inline base_dav/models/dav_collection.py]: why this? According to the spec, this is a timestamp: https://tools.ietf.org/html/rfc6350#section-6.7.4

File: base_dav/radicale/collection.py [hunk 5]
```diff
@@ -120,14 +121,58 @@ def get(self, href):
                         record.write_date
                     ),
                 )
-        record = collection_model.browse(int(components[-1]))
+
+        try:
+            record = collection_model.browse(int(components[-1]))
+        except ValueError:
+            return None
```
Comment (hbrunn) [inline base_dav/radicale/collection.py:128]: will this map to a 404 on the http side? If so, that's better indeed

File: base_dav/radicale/collection.py [hunk 6]
```diff
@@ -120,14 +121,58 @@ def get(self, href):
...
+        if not record_id:
+            record = collection_model.create(data)
+        else:
+            record = collection_model.browse(record_id)
+            record.write(data)
+
+        return Item(
+            self,
+            item=self.collection.to_vobject(record),
+            href=href,
```
Comment (hbrunn) [inline base_dav/radicale/collection.py:160]: don't you need to adapt `href` here for the create case?

**Review by hbrunn — CHANGES_REQUESTED** [review]
thanks, a few minor things

File: base_dav/radicale/collection.py (hunk 6 above)
Comment (fkantelberg) [reply base_dav/radicale/collection.py:160]: You are right that might be a problem if the client doesn't always use list.

File: base_dav/models/dav_collection.py (hunk 4 above)
Comment (fkantelberg) [reply base_dav/models/dav_collection.py]: This was for playing around and I forgot to revert it

**Comment by hbrunn** [comment]
@fkantelberg thanks for the followup. Are you going to change anything or should I merge immediately?

**Comment by hbrunn** [comment]
PS: You might want to squash your commits

**Comment by fkantelberg** [comment]
@hbrunn You can merge

**Pull request merged**